[16.0] improves loyalty_partner_applicability

lmignon opened the pull request:
fixes #238

* [x] includes #267

lmignon committed:
[IMP] loyalty_partner_applicability: Fix partner applicability

Before this commit, the partner applicability was not working according
to the addon description. For programs with multiple rules, rules defined
for specific partners were being applied to any partner as long as the
partner had a rule defined for them. The same applied to the coupon, where
a coupon defined for a specific partner could be used by any partner as long
the partner had a ruele defined for them in the same program where
the coupon was defined.

This commit fixes the issue by ensuring that the partner applicability
is only checked for the partner that the rule or coupon is defined for.
It's done by removing the rules not applicable to the the partner linked
to the sale order when the system is requesting the rules to be applied
to the products in the sale order. When a coupon is applied we also checke
if the partner is linked to the rule at the origin of the coupon. Or to
a rule linked to the loyalty card program.

lmignon committed:
[IMP] loyalty_partner_applicability: Improve the applicability of loyalty programs to partners

This change introduces a new mixin to allow the definition of the applicability
of a loyalty program to a partner. This mixin is used in the loyalty program model
to define the applicability of the program to partners. The mixin is also used
in the loyalty rule model to define the applicability of the rule to partners.

With this change, the applicability of a loyalty program to a partner can be
now defined at the program and rule level.

lmignon committed:
[IMP] loyalty_partner_applicability: Add migration scripts

Rename rule_partners_domain to partner_domain on loyalty.rule

Since the previous implementation was not correct and defacto defined
the applicability of the rule at program level (If a rule was restricted
to a partner, all the rules of the program was available to the partner
even if the other rules where restricted to other partners), this change
add a migration script to aggreate the partner domain of the rules of the
program and set it at the program level. The new domain is the OR of the
domain of the rules of the program.

lmignon committed:
[IMP] loyalty_mass_mailing: Adapt the module to the new way of applying loyalty programs to partners

The partner applicability of a loyalty rule can now be defined at the level of the loyalty program
in addition to the level of the loyalty rule. This commit adapts the module to this new way of
defining the partner applicability of a loyalty program when computing the domain to use for
the mass mailing of a loyalty program.

pedrobaeza commented:
Sorry, I haven't been able yet to check this. In principle, I agree it's weird to select the partner domain at a sublevel. Maybe a trick is to set it at "header", and propagate on the background to each rule, but as said, I'm only talking from the very surface.

lmignon commented:
Thank you for your comment @pedrobaeza . I share the same feeling as yours and can't imagine a valid use case to keep the restriction at the rule level. I keep it at the rule level to ensure the backward compatibility but it comes with a high performance cost. I would really like to remove it from the rule level and only keep it a program level. In this way the restriction could be part of the domain used to search for applicable programs. It will be more efficient.

pilarvargas-tecnativa commented:
@lmignon Can you do a rebase and push again to check in runbot, please?

pilarvargas-tecnativa commented on loyalty_partner_applicability/models/loyalty_partner_applicability_mixin.py:
```diff
@@ -0,0 +1,110 @@
+    _description = "Manage the applicability of loyalty elements to partners"
+
+    partner_ids = fields.Many2many(
+        string="Specific customers",
```
It is best to avoid ‘Specific customers’ because it is ambiguous and does not clearly reflect that the field limits the promotion to certain customers, which may confuse the user about its purpose.

```suggestion
        string="Allowed partners",
```

pilarvargas-tecnativa commented on loyalty_partner_applicability/models/loyalty_partner_applicability_mixin.py:
```diff
@@ -0,0 +1,110 @@
+    partner_ids = fields.Many2many(
+        string="Specific customers",
+        comodel_name="res.partner",
+        help="This field allow to restrict the applicability to a specific set of partners",
```
```suggestion
        help="Only the selected partners will be eligible for this promotion.",
```

pilarvargas-tecnativa commented on loyalty_partner_applicability/models/loyalty_partner_applicability_mixin.py:
```diff
@@ -0,0 +1,110 @@
+    )
+
+    partner_domain = fields.Char(
+        string="Specific customers domain",
```
```suggestion
        string="Allowed partners domain",
```

pilarvargas-tecnativa commented on loyalty_partner_applicability/models/loyalty_partner_applicability_mixin.py:
```diff
@@ -0,0 +1,110 @@
+
+    partner_domain = fields.Char(
+        string="Specific customers domain",
+        help="Applicable to the customers that match the domain",
```
```suggestion
        help="Define the domain to restrict which partners are eligible for this promotion.",
```

pilarvargas-tecnativa commented on loyalty_partner_applicability/models/loyalty_rule.py:
```diff
@@ -1,22 +1,15 @@
-        return res
+class LoyaltyRule(models.Model):
+    _name = "loyalty.rule"
+    _inherit = ["loyalty.rule", "loyalty.partner.applicability.mixin"]
```
If the purpose of refactoring is to apply the rule at the promotion level, I find it a little confusing that a mixin is created and that it is still possible to apply this at the rule level. Ultimately, what is the approach? If it is applied at the promotion level, it would have to be removed at the rule level, as it could define different clients at two different levels, but only one would have priority, that of the rule. This would cause errors when configuring the rule and unexpected results when applying the promotion.

lmignon replied:
@pilarvargas-tecnativa I kept the definition at rule level to not break the existing implementation and don't introduce a change of behavior in used version.

pilarvargas-tecnativa replied:
But it is understood that refactoring brings about this change in behaviour to configure the rules at the promotion level rather than at the rule level. If the current behaviour is to be maintained, then it is not refactoring. It is adding something new and changing the names of what already exists. Correct me if I am wrong.

pilarvargas-tecnativa replied:
I believe that in this case, maintaining both systems seems unnecessary and can be completely simplified. One option would be to replace the old logic by doing an OR in the migration script to manage the different partner domains for each rule. This would maintain data compatibility without complicating the active code.
My recommendation would be to apply this simplification to avoid future problems and facilitate maintenance.

eugenios73 commented:
Hello @lmignon when you solve the comments, tell me and I'll review the functionality.

lmignon committed:
[IMP] loyalty_partner_applicability: Improves field's labels

Co-authored-by: Pilar Vargas <118818446+pilarvargas-tecnativa@users.noreply.github.com>

lmignon commented:
> Hello @lmignon when you solve the comments, tell me and I'll review the functionality.

@eugenios73  comment solved or responded

pilarvargas-tecnativa commented:
Hello @lmignon 

Any news regarding my comments? Would you prefer me to continue with the refactoring as I suggested? Or would you prefer to finish this refactoring yourself?

lmignon commented:
@pilarvargas-tecnativa The problem if we remove the applicability at rule level  could be that you will break existing implementations on v16.

pilarvargas-tecnativa commented:
> @pilarvargas-tecnativa The problem if we remove the applicability at rule level could be that you will break existing implementations on v16.

A script would need to be added to migrate what is established in the rules to the field added in the promotion, combining it with an OR would be sufficient.

The rest is to remove the logic for the rule that has no effect.

lmignon commented:
> > @pilarvargas-tecnativa The problem if we remove the applicability at rule level could be that you will break existing implementations on v16.
> 
> A script would need to be added to migrate what is established in the rules to the field added in the promotion, combining it with an OR would be sufficient.
> 
> The rest is to remove the logic for the rule that has no effect.

@pilarvargas-tecnativa If a program is configured with applicability at rule level, the removing of this functionality will lead to cases where customers could wrongly benefit from loyalties  after the update of this addon on 16.0. I'm really concerned by not introducing unwanted behavior on 16.0...

pilarvargas-tecnativa commented:
> @pilarvargas-tecnativa If a program is configured with applicability at rule level, the removing of this functionality will lead to cases where customers could wrongly benefit from loyalties after the update of this addon on 16.0. I'm really concerned by not introducing unwanted behavior on 16.0...

This is why I mentioned that a migration script must be included to transfer any domains configured in the rules to the new field.

The script must consist of going through the rules, checking for domains, and if there is more than one rule with that domain set, combining them with an OR in the new field.

With this script, the behaviour will be the same when deleting the field from the rules.

lmignon commented:
@pilarvargas-tecnativa If you have 2 rules for the same product with a different amount for different parnters and you combine the the domain a the program level, each partner will benefit of the 2 rules in place of only one.

pilarvargas-tecnativa commented:
> @pilarvargas-tecnativa If you have 2 rules for the same product with a different amount for different parnters and you combine the the domain a the program level, each partner will benefit of the 2 rules in place of only one.

But promotions don't work that way. When there are several rules, as soon as one is met, the promotion is applicable. In other words, if there are two rules for the same product and different partners, it will not discriminate that the rule is met in its entirety, but as soon as the least restrictive condition is met, the rule will be applicable to both partners. You can test this and verify that the domain will function as an OR.  Outside of the partner's applicability, you can configure a rule in which you buy one product (regardless of amount) and another rule in which you buy two units of another product.

When you apply the promotion at the moment when there is one quantity of either of the two products added to different rules, the promotion is applicable, and the same thing happens within the same rule by putting one quantity with one of the two products added to different rules. When you apply the promotion at the moment when there is 1 quantity of either of the two products added to different rules, the promotion is applicable, and the same thing happens within the same rule by putting 1 quantity with X amount; the first condition that is met will apply the rule.

lmignon commented:
I'll change the code, In any case, I will not be affected.

lmignon committed:
[REF] *_loyalty_partler_applicability: No restriction on partner at rule level

At the request of @pilarvargas-tecnativa, removal of the option to limit the applicability of a rule to certain partners only. To simplify the code, the restriction can now only be defined at the programme level.

lmignon commented:
@pilarvargas-tecnativa I made the change, but it seems to me that it still causes a change in behavior (I'm really not convinced it's a good idea since we are loosing flexibility and we don't get a real benefit into the code). If today there is a loyalty program with a coupon for customer X and a rule for customer Y, with the change, after the update of the addon, customer Y can use the coupon and the original UC can no longer be configured.

pilarvargas-tecnativa commented:
> @pilarvargas-tecnativa I made the change, but it seems to me that it still causes a change in behavior (I'm really not convinced it's a good idea since we are loosing flexibility and we don't get a real benefit into the code). If today there is a loyalty program with a coupon for customer X and a rule for customer Y, with the change, after the update of the addon, customer Y can use the coupon and the original UC can no longer be configured.

However, this has not yet been merged, so no one should be using it; there will only be promotions configured at the rule level. When the change is rolled out, the promotion should continue to behave in the same way. In any case, it is not correct to have applicability configured at both levels. And in the hypothetical case that someone has rolled out this PR, if it is more restrictive at the promotion level, then the rule domain should be removed.

pilarvargas-tecnativa commented:
In my local tests:

- I have set up a local database with the above prior to refactoring.
- I have created a promotion with two rules for two different partners and different minimum amounts. The result of applying this promotion is that when the order meets the lowest amount of the rules, it is applicable to either of the two partners.
- I added the PR to my local instance and updated the module to run the migration scripts.
- The domain result is correct; the promotion should be applicable to both partners and therefore the domains have been combined correctly.

On the other hand, the promotion is not being restricted to other partners as expected. The promotion should only be applicable to the two partners in the rules and not to the rest.

pedrobaeza commented:
I'm not sure I'm getting the problem depicted by @lmignon, but he is talking about coupons, not programs. Anyway, I would say once the coupon is issued, it has the partner assigned, so no possibility of being mismatched, isn't it?

pilarvargas-tecnativa commented:
> I'm not sure I'm getting the problem depicted by @lmignon, but he is talking about coupons, not programs. Anyway, I would say once the coupon is issued, it has the partner assigned, so no possibility of being mismatched, isn't it?

Coupons follow a different path.

Gift cards and e-wallets have no rules. For ‘coupon’ promotions that generate coupons, you have to do it manually by setting up the partner, and then there is the promotion that generates a coupon for the next order, which follows the same application rules as all other promotions. If the conditions are met, the coupon for the next order is generated, so the focus is just on migrating the domain correctly, and that’s it.

All that remains is to adjust the application logic so that the promotion is restricted to that domain for the order.

pedrobaeza commented:
So is there a problem or not with this change?

pilarvargas-tecnativa commented on sale_loyalty_partner_applicability/models/sale_order.py:
```diff
@@ -1,77 +1,49 @@
+        # partner domain.
+        # This is needed to check the partner of the order and filter the rules based
+        # on the partner domain.
+        return super()._program_check_compute_points(programs.with_context(order=self))
```
> So is there a problem or not with this change?

The logic would need to be adjusted so that it only applies to the domain, and the scripts would be correct.

pedrobaeza commented:
Then please suggest the needed changes.

pilarvargas-tecnativa reviewed (approved):
I debugged locally and the reason why the restriction was not being applied was due to a problem with the dirty database. I ran the local tests again with everything complete and everything is correct. It is also covered in the added tests.    

LGTM, thanks!

victoralmau reviewed (approved):
Code review OK. Although it may be difficult to review, it will now be more consistent.

lmignon commented:
Thank you to all for your reviews and suggestions. I still think we should have keep the ability to restrict the applicability to partners also at rule level. We could imagine UC where for a same program the reward points for the same product could depends on the partners.

pedrobaeza commented:
@lmignon but AFAIK, that reward part is not tied to the applicability rules, so keeping the partner applicability at that level doesn't improve anything. Can you expand that use case?

lmignon commented:
> @lmignon but AFAIK, that reward part is not tied to the applicability rules, so keeping the partner applicability at that level doesn't improve anything. Can you expand that use case?

@pedrobaeza The number of point rewarded and collected for a product is defined on the rule. 
Imagine the case where for a same program you want to reward customers who have subscribed to a subscription more generously than those without a subscription for the same product.
With the ability to also define the applicability at rule level you can define your program as

Loyalty Program A:
* Rule 1: Product1, reward point 10, domain for Partners with a subscription
* Rule 2: Product1, reward point 5, domain for Partners without subscription

This kind of configuration is no more possible with the last requested change.

pilarvargas-tecnativa commented:
> @pedrobaeza The number of point rewarded and collected for a product is defined on the rule. Imagine the case where for a same program you want to reward customers who have subscribed to a subscription more generously than those without a subscription for the same product. With the ability to also define the applicability at rule level you can define your program as

But that's not really how it works. The rules are combined, not restricted. In your use case:

Loyalty Program A:

- Rule 1: Product1, reward point 10, domain for partner1 
- Rule 2: Product1, reward point 5, domain for partner2 

The result would be the same for both partner1 and partner2 in an order with product1, reward points = rule1+rule2 = 15

lmignon commented:
> > @pedrobaeza The number of point rewarded and collected for a product is defined on the rule. Imagine the case where for a same program you want to reward customers who have subscribed to a subscription more generously than those without a subscription for the same product. With the ability to also define the applicability at rule level you can define your program as
> 
> But that's not really how it works. The rules are combined, not restricted. In your use case:
> 
> Loyalty Program A:
> 
>     * Rule 1: Product1, reward point 10, domain for partner1
> 
>     * Rule 2: Product1, reward point 5, domain for partner2
> 
> 
> The result would be the same for both partner1 and partner2 in an order with product1, reward points = rule1+rule2 = 15



> > @pedrobaeza The number of point rewarded and collected for a product is defined on the rule. Imagine the case where for a same program you want to reward customers who have subscribed to a subscription more generously than those without a subscription for the same product. With the ability to also define the applicability at rule level you can define your program as
> 
> But that's not really how it works. The rules are combined, not restricted. In your use case:
> 
> Loyalty Program A:
> 
>     * Rule 1: Product1, reward point 10, domain for partner1
> 
>     * Rule 2: Product1, reward point 5, domain for partner2
> 
> 
> The result would be the same for both partner1 and partner2 in an order with product1, reward points = rule1+rule2 = 15

With my code, the rule1 will not be applied on order from partner2 and vice versa.

pedrobaeza commented:
But that means that this module is also changing other part of the `loyalty` behavior, not the partner applicability, so that should go in another module, not here. I think this module should be kept simple, and let the lines part on an extra module on top of it. It was your initial argumentation for the refactorization: simplify the code and the user experience. In current form, I think it's the contrary, having more code and possible combinations that leds to user confusion.

lmignon commented:
@pedrobaeza I made all the changes requested even the one I'm not convinced it should be done. I improved the way you could define your restriction on partners, and the use of the mixin before the last change makes the code easy to understand and to maintain. I also added a lot of tests for all the UCs. The day I need to allows to restrict partner at rule level, the best way to do it is to add a mixin as it was done to share the logic between the program and the rule... I hope this PR should at least be merged as is.

pedrobaeza commented:
OK, don't worry, I don't argue more, so let's merge it.

/ocabot merge major
